File: Terrain/CFigure.cpp
/*-----------------------------------------------------------------------------

  Figure.cpp

-------------------------------------------------------------------------------

  Animated models.

-----------------------------------------------------------------------------*/

#include "stdafx.h"

#include "console.h"
#include "cfigure.h"
#include "cg.h"
#include "file.h"

#define NEWLINE   "\n"

static void clean_chars (char* target, char* chars)
{

  unsigned    i;
  char*       c;

  for (i = 0; i < strlen (chars); i++) {
    while (c = strchr (target, chars[i]))
      *c = ' ';
  }

}

CFigure::CFigure ()
{

  Clear ();

}

void CFigure::Clear ()
{

  unsigned    i;

  for (i = 0; i < BONE_COUNT; i++) 
    _bone_index[i] = BONE_INVALID;
  _unknown_count = 0;
  _skin_static.Clear ();
  _skin_deform.Clear ();
  _skin_render.Clear ();
  _bone.clear ();


}

void CFigure::Animate (CAnim* anim, float delta)
{

  AnimJoint*    aj;
  
  if (delta > 1.0f)
    delta -= (float)((int)delta);
  aj = anim->GetFrame (delta);
  for (unsigned i = 0; i < anim->Joints (); i++) 
     RotateBone (aj[i].id, aj[i].rotation);



}

//We take a string and turn it into a BoneId, using unknowns as needed
BoneId CFigure::IdentifyBone (char* name)
{
  
  BoneId    bid;

  bid = CAnim::BoneFromString (name);
  //If CAnim couldn't make sense of the name, or if that id is already in use...
  if (bid == BONE_INVALID || _bone_index[bid] != BONE_INVALID) {
    ConsoleLog ("Couldn't id Bone '%s'.", name);
    bid = (BoneId)(BONE_UNKNOWN0 + _unknown_count);
    _unknown_count++;
  }
  return bid;

}

void CFigure::RotateBone (BoneId id, GLvector angle)
{

  if (_bone_index[id] != BONE_INVALID)
    _bone[_bone_index[id]]._rotation = angle;

}

void CFigure::RotatePoints (unsigned id, GLvector offset, GLmatrix m)
{

  Bone*       b;
  unsigned    i;
  unsigned    index;

  
  b = &_bone[_bone_index[id]];
  for (i = 0; i < b->_vertex_weights.size (); i++) {
    index = b->_vertex_weights[i]._index;
    _skin_render._vertex[index] = glMatrixTransformPoint (m, _skin_render._vertex[index] - offset) + offset;
    /*
    from = _skin_render._vertex[index] - offset;
    to = glMatrixTransformPoint (m, from);
    //movement = movement - _skin_static._vertex[index]; 
    _skin_render._vertex[index] = glVectorInterpolate (from, to, b->_vertex_weights[i]._weight) + offset;
    */
  }

}

void CFigure::RotateHierarchy (unsigned id, GLvector offset, GLmatrix m)
{

  Bone*       b;
  unsigned    i;

  b = &_bone[_bone_index[id]];
  b->_position = glMatrixTransformPoint (m, b->_position - offset) + offset;
  RotatePoints (id, offset, m);
  for (i = 0; i < b->_children.size (); i++) {
    if (b->_children[i])
      RotateHierarchy (b->_children[i], offset, m);
  }

}

void CFigure::Update ()
{

  unsigned    i;
  unsigned    c;
  GLmatrix    m;
  Bone*       b;

  _skin_render = _skin_deform;
  for (i = 1; i < _bone.size (); i++) 
    _bone[i]._position = _bone[i]._origin;

  vector<Bone>::reverse_iterator rit;
  for (rit = _bone.rbegin(); rit < _bone.rend(); ++rit) {
    b = &(*rit);
    if (b->_rotation == glVector (0.0f, 0.0f, 0.0f))
      continue;
    if (b->_id == BONE_ROOT)
      m.Identity ();

    m.Identity ();
    m.Rotate (b->_rotation.x, 1.0f, 0.0f, 0.0f);
    m.Rotate (b->_rotation.y, 0.0f, 1.0f, 0.0f);
    m.Rotate (b->_rotation.z, 0.0f, 0.0f, 1.0f);
    RotatePoints (b->_id, b->_position, m);
    for (c = 0; c < b->_children.size (); c++) 
      RotateHierarchy (b->_children[c], b->_position, m);
  }
  
}

void CFigure::PushWeight (unsigned id, unsigned index, float weight)
{

  BWeight   bw;

  bw._index = index;
  bw._weight = weight;  
  _bone[_bone_index[id]]._vertex_weights.push_back (bw);

}

void CFigure::PushBone (BoneId id, unsigned parent, GLvector pos)
{

  Bone    b;

  _bone_index[id] = _bone.size ();
  b._id = (BoneId)id;
  b._id_parent = (BoneId)parent;
  b._position = pos;
  b._origin = pos;
  b._rotation = glVector (0.0f, 0.0f, 0.0f);
  b._children.clear ();
  b._color = glRgbaUnique (id + 1);
  _bone.push_back (b);
  _bone[_bone_index[parent]]._children.push_back (id);

}

void CFigure::BoneInflate (BoneId id, float distance, bool do_children)
{

  Bone*       b;
  unsigned    i;
  unsigned    c;
  unsigned    index;

  b = &_bone[_bone_index[id]];
  for (i = 0; i < b->_vertex_weights.size (); i++) {
    index = b->_vertex_weights[i]._index;
    _skin_deform._vertex[index] = _skin_static._vertex[index] + _skin_static._normal[index] * distance;
  }
  if (!do_children)
    return;
  for (c = 0; c < b->_children.size (); c++) 
    BoneInflate ((BoneId)b->_children[c], distance, do_children);


}

void CFigure::RotationSet (GLvector rot)
{

  _bone[_bone_index[BONE_ROOT]]._rotation = rot;

}

void CFigure::Render ()
{
  
  glColor3f (1,1,1);
  glPushMatrix ();
  glTranslatef (_position.x, _position.y, _position.z); 
  CgUpdateMatrix ();
  _skin_render.Render ();
  glPopMatrix ();
  CgUpdateMatrix ();
  
}

void CFigure::RenderSkeleton ()
{

  unsigned    i;
  unsigned    parent;

  glLineWidth (12.0f);
  glPushMatrix ();
  glTranslatef (_position.x, _position.y, _position.z); 
  CgUpdateMatrix ();
  glDisable (GL_DEPTH_TEST);
  glDisable (GL_TEXTURE_2D);
  glDisable (GL_LIGHTING);
  for (i = 0; i < _bone.size (); i++) {
    parent = _bone_index[_bone[i]._id_parent];
    if (!parent)
      continue;
    glColor3fv (&_bone[i]._color.red);
    glBegin (GL_LINES);
    GLvector p = _bone[i]._position;
    glVertex3fv (&_bone[i]._position.x);
    glVertex3fv (&_bone[parent]._position.x);
    glEnd ();
  }
  glLineWidth (1.0f);
  glEnable (GL_LIGHTING);
  glEnable (GL_TEXTURE_2D);
  glEnable (GL_DEPTH_TEST);
  glPopMatrix ();
  CgUpdateMatrix ();

}

void CFigure::Prepare ()
{

  _skin_deform = _skin_static;

}

bool CFigure::LoadX (char* filename)
{

  FileXLoad (filename, this);
  Prepare ();
  return true;

}

File: Terrain/CFigure.cs
/*-----------------------------------------------------------------------------
  Figure.cs
  Animated models.
-----------------------------------------------------------------------------*/

using System.Collections.Generic;

using OpenTK;
using OpenTK.Graphics;

namespace Frontier {
	struct BWeight {
		public int   index;
		public float weight;
	};

	struct PWeight {
		public BoneId bone;
		public float  weight;
	};

	struct Bone {
		public BoneId					id, idParent;
		public Vector3				origin, position, rotation;
		public Color4					color;
		public List<int>			children;
		public List<BWeight>	vertex_weights;
		public Matrix4				matrix;
	};

	class CFigure {
		#region Member fields, constants and properties
		private const char NEWLINE = '\n';

		private List<Bone> mBones;
		private Vector3 mPosition, mRotation;
		private int[] mBoneIndices = new int[BONE_COUNT];
		private int             mUnknownCount;

		private Mesh
			mSkinStatic,		//The original, "read only"
			mSkinDeform,		//Altered
			mSkinRender;		//Updated every frame

		public Vector3 Position { get; set; }
		public Mesh Skin { get { return mSkinStatic; } }
		#endregion

		#region Public methods
		public CFigure() { Clear(); }

		public void Clear() {
			for (int i = 0; i < BONE_COUNT; i++)
				mBoneIndices[i] = BONE_INVALID;
			mUnknownCount = 0;
			mSkinStatic.Clear();
			mSkinDeform.Clear();
			mSkinRender.Clear();
			mBones.Clear();
		}

		public void Animate(CAnim anim, float delta) {
			List<AnimJoint> aj;

			if (delta > 1.0f)
				delta -= (float) ((int) delta);
			aj = anim.GetFrame(delta);
			for (int i = 0; i < anim.JointCount; i++)
				RotateBone(aj[i].id, aj[i].rotation);
		}

		//We take a string and turn it into a BoneId, using unknowns as needed
		public BoneId IdentifyBone (string name) {
			BoneId    bid;

			bid = CAnim.BoneFromString (name);
			//If CAnim couldn't make sense of the name, or if that id is already in use...
			if (bid == BONE_INVALID || mBoneIndices[bid] != BONE_INVALID) {
				//ConsoleLog ("Couldn't id Bone '%s'.", name);
				bid = (BoneId)(BONE_UNKNOWN0 + mUnknownCount);
				mUnknownCount++;
			}
			return bid;
		}

		public void RotateBone(BoneId id, Vector3 angle) {
			if (mBoneIndices[id] != BONE_INVALID)
				mBones[mBoneIndices[id]].mRotation = angle;
		}

		public void Update() {
			int    i;
			int    c;
			Matrix4    m;
			Bone*       b;
			List<Bone>.reverse_iterator rit;

			mSkinRender = mSkinDeform;
			for (i = 1; i < mBones.size (); i++) 
				mBones[i].mPosition = mBones[i].origin;
			for (rit = mBones.rbegin(); rit < mBones.rend(); ++rit) {
				b = &(*rit);
				if (b.mRotation == Vector3 (0.0f, 0.0f, 0.0f))
					continue;
				m.Identity ();
				m.Rotate (b.mRotation.x, 1.0f, 0.0f, 0.0f);
				m.Rotate (b.mRotation.z, 0.0f, 0.0f, 1.0f);
				m.Rotate (b.mRotation.y, 0.0f, 1.0f, 0.0f);
				RotatePoints (b._id, b.mPosition, m);
				for (c = 0; c < b._children.size (); c++) {
					//Root is self-parent, but shouldn't rotate self!
					if (b._children[c]) 
						RotateHierarchy (b._children[c], b.mPosition, m);
				}
			}
		}

		public void PushWeight(int id, int index, float weight) {
			BWeight   bw;

			bw.index = index;
			bw.weight = weight;
			mBones[mBoneIndices[id]].vertex_weights.push_back(bw);
		}

		public void PushBone(BoneId id, int parent, Vector3 pos) {
			Bone    b;

			mBoneIndices[(int) id] = mBones.Count;
			b.id = (BoneId) id;
			b.idParent = (BoneId) parent;
			b.position = pos;
			b.origin = pos;
			b.rotation = Vector3.Zero;
			b.children.Clear();
			b.color = glRgbaUnique(id + 1);
			mBones.Add(b);
			mBones[mBoneIndices[parent]].children.Add((int) id);
		}

		public void BoneInflate(BoneId id, float distance, bool do_children) {
			Bone b = mBones[mBoneIndices[(int) id]];
			for (int i = 0; i < b.vertex_weights.Count; i++) {
				int index = b.vertex_weights[i].index;
				mSkinDeform.vertices[index] = mSkinStatic.vertices[index] + mSkinStatic.normals[index] * distance;
			}
			if (!do_children)
				return;
			for (int c = 0; c < b.children.Count; c++)
				BoneInflate((BoneId) b.children[c], distance, do_children);
		}

		public void RotationSet(Vector3 rot) { mBones[mBoneIndices[BONE_ROOT]].mRotation = rot; }

		public void Render() {
			glColor3f(1, 1, 1);
			glPushMatrix();
			CgSetOffset(mPosition);
			glTranslatef(mPosition.x, mPosition.y, mPosition.z);
			CgUpdateMatrix();
			mSkinRender.Render();
			CgSetOffset(Vector3(0, 0, 0));
			glPopMatrix();
			CgUpdateMatrix();
		}

		public void RenderSkeleton() {
			int    i;
			int    parent;

			glLineWidth(12.0f);
			glPushMatrix();
			glTranslatef(mPosition.x, mPosition.y, mPosition.z);
			CgUpdateMatrix();
			glDisable(GL_DEPTH_TEST);
			glDisable(GL_TEXTURE_2D);
			glDisable(GL_LIGHTING);
			for (i = 0; i < mBones.size(); i++) {
				parent = mBoneIndices[mBones[i].idParent];
				if (!parent)
					continue;
				glColor3fv(&mBones[i].color.red);
				glBegin(GL_LINES);
				Vector3 p = mBones[i].mPosition;
				glVertex3fv(&mBones[i].mPosition.x);
				glVertex3fv(&mBones[parent].mPosition.x);
				glEnd();
			}
			glLineWidth(1.0f);
			glEnable(GL_LIGHTING);
			glEnable(GL_TEXTURE_2D);
			glEnable(GL_DEPTH_TEST);
			glPopMatrix();
			CgUpdateMatrix();
		}

		public void Prepare() { mSkinDeform = mSkinStatic; }

		public bool LoadX(string filename) {
			FileXLoad(filename, this);
			Prepare();
			return true;
		}
		#endregion

		#region Private methods
		private void RotatePoints(int id, Vector3 offset, Matrix4 m) {
			Bone*       b;
			int    i;
			int    index;

			b = &mBones[mBoneIndices[id]];
			for (i = 0; i < b._vertex_weights.size(); i++) {
				index = b._vertex_weights[i]._index;
				mSkinRender._vertex[index] = Matrix4TransformPoint(m, mSkinRender._vertex[index] - offset) + offset;
				/*
				from = mSkinRender._vertex[index] - offset;
				to = Matrix4TransformPoint (m, from);
				//movement = movement - mSkinStatic._vertex[index]; 
				mSkinRender._vertex[index] = Vector3Interpolate (from, to, b.vertex_weights[i].weight) + offset;
				*/
			}
		}

		private void RotateHierarchy(int id, Vector3 offset, Matrix4 m) {
			Bone*       b;
			int    i;

			b = &mBones[mBoneIndices[id]];
			b.mPosition = Matrix4TransformPoint(m, b.mPosition - offset) + offset;
			RotatePoints(id, offset, m);
			for (i = 0; i < b._children.size(); i++) {
				if (b._children[i])
					RotateHierarchy(b._children[i], offset, m);
			}
		}

		private static void clean_chars(char* target, char* chars) {
			char* c;

			for (int i = 0; i < strlen(chars); i++) {
				while (c = strchr(target, chars[i]))
					*c = ' ';
			}
		}
		#endregion
	};
}